lib/ansible/modules/cloud/amazon/iam_saml_federation.py

```diff
@@ -0,0 +1,267 @@
+            description: The XML metadata document that includes information about an identity provider.
+            type: str
+            returned: present
+        create_date:
+            description: The date and time when the SAML provider was created in ISO 8601 date-time format.
+            type: str
+            returned: present
+            sample: "2017-02-08T04:36:28+00:00"
+        expire_date:
+            description: The expiration date and time for the SAML provider in ISO 8601 date-time format.
+            type: str
+            returned: present
+            sample: "2017-02-08T04:36:28+00:00"
+'''
+
+try:
+    import boto3
+    import botocore.exceptions
+
+    HAS_BOTO3 = True
+except ImportError:
+    HAS_BOTO3 = False
+
+from ansible.module_utils.aws.core import AnsibleAWSModule
+from ansible.module_utils.ec2 import AWSRetry, aws_common_argument_spec, get_aws_connection_info, boto3_conn
+
+
+class SAMLProviderManager:
+    """Handles SAML Identity Provider configuration"""
+
+    def __init__(self, module, **aws_connection_params):
+        self.module = module
+        self.aws_connection_params = aws_connection_params
+
+        try:
+            self.conn = boto3_conn(module, conn_type='client', resource='iam', **self.aws_connection_params)
+        except botocore.exceptions.ClientError as e:
+            self.module.fail_json(msg=str(e))
+
+    # use retry decorator for boto3 calls
+    @AWSRetry.backoff(tries=3, delay=5)
+    def _list_saml_providers(self):
+        return self.conn.list_saml_providers()
+
+    @AWSRetry.backoff(tries=3, delay=5)
+    def _get_saml_provider(self, arn):
+        return self.conn.get_saml_provider(SAMLProviderArn=arn)
+
+    @AWSRetry.backoff(tries=3, delay=5)
+    def _update_saml_provider(self, arn, metadata):
+        return self.conn.update_saml_provider(SAMLProviderArn=arn, SAMLMetadataDocument=metadata)
+
+    @AWSRetry.backoff(tries=3, delay=5)
+    def _create_saml_provider(self, metadata, name):
+        return self.conn.create_saml_provider(SAMLMetadataDocument=metadata, Name=name)
+
+    @AWSRetry.backoff(tries=3, delay=5)
+    def _delete_saml_provider(self, arn):
+        return self.conn.delete_saml_provider(SAMLProviderArn=arn)
+
+    def _get_provider_arn(self, name):
+        providers = self._list_saml_providers()
+        for p in providers['SAMLProviderList']:
+            provider_name = p['Arn'].split('/', 1)[1]
+            if name == provider_name:
+                return p['Arn']
+
+        return None
+
+    def create_or_update_saml_provider(self, name, metadata):
+        if not metadata:
+            self.module.fail_json(msg="saml_metadata_document must be defined for present state")
+
+        res = {'changed': False}
+        try:
+            arn = self._get_provider_arn(name)
+        except (botocore.exceptions.ValidationError, botocore.exceptions.ClientError) as e:
+            res['msg'] = str(e)
+            self.module.fail_json(**res)
```

Comment:
This (and the other exception handling) should use the standard convenience function, which will format the exception for you:

```self.module.fail_json_aws(e, msg='Error getting provider ARN for {0}'.format(name))```